ES doesn't start when there are empty cgroup controller names in '/proc/self/cgroup'

<!--
GitHub is reserved for bug reports and feature requests. The best place
to ask a general question is at the Elastic Discourse forums at
https://discuss.elastic.co. If you are in fact posting a bug report or
a feature request, please include one and only one of the below blocks
in your new issue. Note that whether you're filing a bug report or a
feature request, ensure that your submission is for an
[OS that we support](https://www.elastic.co/support/matrix#show_os).
Bug reports on an OS that we do not support or feature requests
specific to an OS that we do not support will be closed.
-->

<!--
If you are filing a bug report, please remove the below feature
request block and provide responses for all of the below items.
-->

**Elasticsearch version**: 5.2.2

**Plugins installed**: []

**JVM version**: 8u122

**OS version**: Linux

**Description of the problem including expected versus actual behavior**:
The OS Probe Regex fails if there is a cgroup entry with no controller and crashes. Example ``/proc/self/cgroup`` (see last line):
<pre>
8:net_cls:/
7:devices:/user.slice
6:pids:/user.slice/user-1000.slice/session-3.scope
5:blkio:/
4:freezer:/
3:memory:/
2:cpu,cpuacct:/
1:cpuset:/
0::/user.slice/user-1000.slice/session-3.scope
</pre>

Related issue: https://github.com/elastic/elasticsearch/issues/23218

**Steps to reproduce**:
 1.
 2.
 3.

**Provide logs (if relevant)**:
<pre>
Mar 03 12:43:00 elasticsearch systemd[1]: elasticsearch.service: Main process exited, code=exited, status=1/FAILURE
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         ... 6 more
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:121) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:333) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:241) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.bootstrap.Bootstrap$6.<init>(Bootstrap.java:241) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.node.Node.<init>(Node.java:232) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.node.Node.<init>(Node.java:345) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.monitor.MonitorService.<init>(MonitorService.java:45) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.monitor.os.OsService.<init>(OsService.java:45) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.monitor.os.OsProbe.osStats(OsProbe.java:466) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.monitor.os.OsProbe.getCgroup(OsProbe.java:414) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.monitor.os.OsProbe.getControlGroups(OsProbe.java:216) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at java.util.regex.Matcher.group(Matcher.java:536) ~[?:1.8.0_121]
Mar 03 12:43:00 elasticsearch elasticsearch[641]: Caused by: java.lang.IllegalStateException: No match found
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:82) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:89) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.cli.Command.main(Command.java:88) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:122) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.cli.SettingCommand.execute(SettingCommand.java:54) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:112) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]:         at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:125) ~[elasticsearch-5.2.2.jar:5.2.2]
Mar 03 12:43:00 elasticsearch elasticsearch[641]: org.elasticsearch.bootstrap.StartupException: java.lang.IllegalStateException: No match found
Mar 03 12:43:00 elasticsearch elasticsearch[641]: [2017-03-03T12:43:00,027][WARN ][org.elasticsearch.bootstrap.ElasticsearchUncaughtExceptionHandler] uncaught exception in thread [main]
Mar 03 12:42:59 elasticsearch elasticsearch[641]: [2017-03-03T12:42:59,000][WARN ][org.elasticsearch.deprecation.script.groovy.GroovyScriptEngineService] [groovy] scripts are deprecated, use [painless] scripts instead
Mar 03 12:42:58 elasticsearch elasticsearch[641]: [2017-03-03T12:42:58,412][INFO ][org.elasticsearch.plugins.PluginsService] no plugins loaded
</pre>
<!--
If you are filing a feature request, please remove the above bug
report block and provide responses for all of the below items.
-->

**Describe the feature**:
Make the regex more robust.Changing the ``+`` to a ``*`` in the failing regex for the part matching the cgroup controller name should do the trick. I would make a PR, but I am not willing to sign the CLA.

--- a/core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java
+++ b/core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java
@@ -187,9 +187,6 @@ public class OsProbe {
         return lines.get(0);
     }
 
-    // pattern for lines in /proc/self/cgroup
-    private static final Pattern CONTROL_GROUP_PATTERN = Pattern.compile("\\d+:([^:]+):(/.*)");
-
     // this property is to support a hack to workaround an issue with Docker containers mounting the cgroups hierarchy inconsistently with
     // respect to /proc/self/cgroup; for Docker containers this should be set to "/"
     private static final String CONTROL_GROUPS_HIERARCHY_OVERRIDE = System.getProperty("es.cgroups.hierarchy.override");
@@ -205,23 +202,29 @@ public class OsProbe {
         final List<String> lines = readProcSelfCgroup();
         final Map<String, String> controllerMap = new HashMap<>();
         for (final String line : lines) {
-            final Matcher matcher = CONTROL_GROUP_PATTERN.matcher(line);
-            // Matcher#matches must be invoked as matching is lazy; this can not happen in an assert as assertions might not be enabled
-            final boolean matches = matcher.matches();
-            assert matches : line;
-            // at this point we have captured the subsystems and the control group
-            final String[] controllers = matcher.group(1).split(",");
+            /*
+             * The virtual file /proc/self/cgroup lists the control groups that the Elasticsearch process is a member of. Each line contains
+             * three colon-separated fields of the form hierarchy-ID:subsystem-list:cgroup-path. For cgroups version 1 hierarchies, the
+             * subsystem-list is a comma-separated list of subsystems. The subsystem-list can be empty if the hierarchy represents a cgroups
+             * version 2 hierarchy. For cgroups version 1
+             */
+            final String[] fields = line.split(":");
+            assert fields.length == 3;
+            final String[] controllers = fields[1].split(",");
             for (final String controller : controllers) {
+                final String controlGroupPath;
                 if (CONTROL_GROUPS_HIERARCHY_OVERRIDE != null) {
                     /*
                      * Docker violates the relationship between /proc/self/cgroup and the /sys/fs/cgroup hierarchy. It's possible that this
                      * will be fixed in future versions of Docker with cgroup namespaces, but this requires modern kernels. Thus, we provide
                      * an undocumented hack for overriding the control group path. Do not rely on this hack, it will be removed.
                      */
-                    controllerMap.put(controller, CONTROL_GROUPS_HIERARCHY_OVERRIDE);
+                    controlGroupPath = CONTROL_GROUPS_HIERARCHY_OVERRIDE;
                 } else {
-                    controllerMap.put(controller, matcher.group(2));
+                    controlGroupPath = fields[2];
                 }
+                final String previous = controllerMap.put(controller, controlGroupPath);
+                assert previous == null;
             }
         }
         return controllerMap;

--- a/core/src/test/java/org/elasticsearch/monitor/os/OsProbeTests.java
+++ b/core/src/test/java/org/elasticsearch/monitor/os/OsProbeTests.java
@@ -155,16 +155,17 @@ public class OsProbeTests extends ESTestCase {
             @Override
             List<String> readProcSelfCgroup() {
                 return Arrays.asList(
-                    "10:freezer:/",
-                    "9:net_cls,net_prio:/",
-                    "8:pids:/",
-                    "7:blkio:/",
-                    "6:memory:/",
-                    "5:devices:/user.slice",
-                    "4:hugetlb:/",
-                    "3:perf_event:/",
-                    "2:cpu,cpuacct,cpuset:/" + hierarchy,
-                    "1:name=systemd:/user.slice/user-1000.slice/session-2359.scope");
+                        "10:freezer:/",
+                        "9:net_cls,net_prio:/",
+                        "8:pids:/",
+                        "7:blkio:/",
+                        "6:memory:/",
+                        "5:devices:/user.slice",
+                        "4:hugetlb:/",
+                        "3:perf_event:/",
+                        "2:cpu,cpuacct,cpuset:/" + hierarchy,
+                        "1:name=systemd:/user.slice/user-1000.slice/session-2359.scope",
+                        "0::/cgroup2");
             }
 
             @Override

